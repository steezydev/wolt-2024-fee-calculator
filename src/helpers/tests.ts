const localStorageMock = (function () {
    let store: Record<string, string> = {};
    return {
        getItem(key: string) {
            return store[key] || null;
        },
        setItem(key: string, value: string) {
            store[key] = value.toString();
        },
        clear() {
            store = {};
        },
    };
})();

const matchMediaMock = jest.fn().mockImplementation((query) => ({
    matches: query === '(prefers-color-scheme: dark)',
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
}));

export const mockContext = () => {
    beforeAll(() => {
        Object.defineProperty(window, 'localStorage', {
            value: {
                getItem: jest.fn().mockImplementation((key: string) => {
                    return localStorageMock.getItem(key);
                }),
                setItem: jest
                    .fn()
                    .mockImplementation((key: string, value: string) => {
                        localStorageMock.setItem(key, value);
                    }),
                clear: localStorageMock.clear,
            },
            writable: true,
        });

        window.matchMedia = matchMediaMock;
    });

    beforeEach(() => {
        (window.localStorage.getItem as jest.Mock).mockClear();
        (window.localStorage.setItem as jest.Mock).mockClear();
        matchMediaMock.mockClear();
    });
};

export const setupFakeDate = () => {
    beforeEach(() => {
        jest.useFakeTimers();
        jest.setSystemTime(new Date('2024-01-01'));
    });

    afterEach(() => {
        jest.useRealTimers();
    });
};

export const mockScrollTo = () => {
    beforeEach(() => {
        Element.prototype.scrollTo = jest.fn();
    });

    afterEach(() => {
        (Element.prototype.scrollTo as jest.Mock).mockRestore();
    });
};
