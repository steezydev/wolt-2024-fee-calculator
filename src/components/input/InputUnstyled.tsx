import { classNames } from '@/helpers/classNames';
import { InputProps } from '@/types/props/InputProps';
import React, { useState } from 'react';

const InputUnstyled = React.forwardRef<HTMLInputElement, InputProps>(
  (
    {
      id,
      ariaLabel,
      type,
      name,
      value,
      onChange,
      onFocus,
      onBlur,
      onInput,
      validate,
      placeholder,
      min,
      max,
      autoFocus = false,
      disabled = false,
      required = false,
      className,
    },
    ref
  ) => {
    const [isValid, setIsValid] = useState(true);

    //TODO: Make validation
    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const newValue = event.target.value;
      onChange && onChange(newValue);
      if (validate) {
        setIsValid(validate(newValue));
      }
    };

    return (
      <input
        ref={ref}
        type={type}
        id={id}
        name={name}
        aria-label={ariaLabel}
        value={value}
        onChange={handleChange}
        onFocus={onFocus}
        onBlur={onBlur}
        disabled={disabled}
        required={required}
        placeholder={placeholder}
        className={classNames(
          '[appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none',
          className
        )}
        onInput={onInput}
        autoFocus={autoFocus}
        aria-invalid={!isValid}
        data-test-id={id}
        min={min}
        max={max}
      />
    );
  }
);

export default InputUnstyled;
