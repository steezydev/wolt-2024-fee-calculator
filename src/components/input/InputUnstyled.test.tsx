import { inputRenderTests } from './Input.tests';
import InputUnstyled from './InputUnstyled';

inputRenderTests('InputUnstyled', InputUnstyled);
