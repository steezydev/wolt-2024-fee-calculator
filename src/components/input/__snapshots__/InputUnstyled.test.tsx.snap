// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputUnstyled when asked to render autofocused renders autofocused 1`] = `
<input
  aria-invalid={false}
  aria-label="Test label"
  autoFocus={true}
  className="[appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
  data-test-id="test-input"
  disabled={false}
  id="test-input"
  name="test-input"
  onChange={[Function]}
  required={false}
  type="text"
/>
`;

exports[`InputUnstyled when asked to render disabled renders disabled 1`] = `
<input
  aria-invalid={false}
  aria-label="Test label"
  autoFocus={false}
  className="[appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
  data-test-id="test-input"
  disabled={true}
  id="test-input"
  name="test-input"
  onChange={[Function]}
  required={false}
  type="text"
/>
`;

exports[`InputUnstyled when asked to render renders 1`] = `
<input
  aria-invalid={false}
  aria-label="Test label"
  autoFocus={false}
  className="[appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
  data-test-id="test-input"
  disabled={false}
  id="test-input"
  name="test-input"
  onChange={[Function]}
  required={false}
  type="text"
/>
`;

exports[`InputUnstyled when asked to render required renders required 1`] = `
<input
  aria-invalid={false}
  aria-label="Test label"
  autoFocus={false}
  className="[appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
  data-test-id="test-input"
  disabled={false}
  id="test-input"
  name="test-input"
  onChange={[Function]}
  required={true}
  type="text"
/>
`;

exports[`InputUnstyled when asked to render with placeholder renders with placeholder 1`] = `
<input
  aria-invalid={false}
  aria-label="Test label"
  autoFocus={false}
  className="[appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
  data-test-id="test-input"
  disabled={false}
  id="test-input"
  name="test-input"
  onChange={[Function]}
  placeholder="Test placeholder"
  required={false}
  type="text"
/>
`;

exports[`InputUnstyled when asked to render with value renders with placeholder 1`] = `
<input
  aria-invalid={false}
  aria-label="Test label"
  autoFocus={false}
  className="[appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
  data-test-id="test-input"
  disabled={false}
  id="test-input"
  name="test-input"
  onChange={[Function]}
  required={false}
  type="text"
  value="Test value"
/>
`;
